### PR TITLE
Add pafpn

### PR DESCRIPTION
**Pull request by yhcao6**

### CODE REVIEW

File: configs/pafpn/faster_rcnn_r50_fpn_1x_coco.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+_base_ = [
+    '../_base_/models/faster_rcnn_r50_fpn.py',
+    '../_base_/datasets/coco_detection.py',
+    '../_base_/schedules/schedule_1x.py', '../_base_/default_runtime.py'
+]
```
Comment (xvjiarui) [inline configs/pafpn/faster_rcnn_r50_fpn_1x_coco.py:5]: We may inherit from `../faster_rcnn/faster_rcnn_r50_fpn_1x_coco.py`. 
Also, rename this file to `faster_rcnn_r50_pafpn_1x_coco.py`.

File: mmdet/models/necks/pafpn.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+import torch.nn.functional as F
+from mmcv.cnn import xavier_init
+
+from mmdet.core import auto_fp16
+from mmdet.ops import ConvModule
+from ..registry import NECKS
+
+
+@NECKS.register_module
+class PAFPN(nn.Module):
```
Comment (xvjiarui) [inline mmdet/models/necks/pafpn.py:11]: Could we inherit from FPN to reuse the FPN part of the model? It should reduce some duplicate code.

File: mmdet/models/necks/pafpn.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+@NECKS.register_module
+class PAFPN(FPN):
+    """Path Aggregation Network for Instance Segmentation.
+
+    This is an implementation of the PAFPN in Path Aggregation Network
+    (https://arxiv.org/abs/1803.01534)
+
+    Args:
+        in_channels (List[int]): number of input channels per scale
+
```
Comment (hellock) [inline mmdet/models/necks/pafpn.py:19]: - No need to add a blank line here.
- Use upper case for the first character, e.g., Number.

File: mmdet/models/necks/pafpn.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        # build top-down path
+        used_backbone_levels = len(laterals)
+        for i in range(used_backbone_levels - 1, 0, -1):
+            prev_shape = laterals[i - 1].shape[2:]
+            laterals[i - 1] += F.interpolate(
+                laterals[i], size=prev_shape, mode='nearest')
+
+        # build outputs
+        # part 1: from original levels
+        outs = [
```
Comment (hellock) [inline mmdet/models/necks/pafpn.py:90]: `inter_outs`

File: mmdet/models/necks/pafpn.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            self.fpn_convs[i](laterals[i]) for i in range(used_backbone_levels)
+        ]
+
+        # part 2: add pafpn
```
Comment (hellock) [inline mmdet/models/necks/pafpn.py:94]:
```python
# part 2: add bottom-up path
```

File: mmdet/models/necks/pafpn.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        for i in range(0, used_backbone_levels - 1):
+            outs[i + 1] += self.lateral_deconvs[i](outs[i])
+
+        paouts = []
```
Comment (hellock) [inline mmdet/models/necks/pafpn.py:98]: `outs`

File: mmdet/models/necks/pafpn.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+                 conv_cfg=None,
+                 norm_cfg=None,
+                 act_cfg=None):
+        super(PAFPN,
+              self).__init__(in_channels, out_channels, num_outs, start_level,
+                             end_level, add_extra_convs, extra_convs_on_inputs,
+                             relu_before_extra_convs, no_norm_on_lateral,
+                             conv_cfg, norm_cfg, act_cfg)
+        # add extra bottom up pathway
+        self.lateral_deconvs = nn.ModuleList()
```
Comment (hellock) [inline mmdet/models/necks/pafpn.py:58]: The variable name is misleading. Use `downsample_convs` instead.

**Pull request merged**